database/src/main/java/com/firebase/ui/database/paging/FirebaseDataSource.java

```diff
@@ -0,0 +1,239 @@
+    private static final String DETAILS_DATABASE_NOT_FOUND = "Database Children Not Found in the specified child path. Please specify correct child path/reference";
+
+    private Runnable mRetryRunnable;
+
+    public static class Factory extends DataSource.Factory<String, DataSnapshot> {
+
+        private final Query mQuery;
+
+        public Factory(@NonNull Query query) {
+            mQuery = query;
+        }
+
+        @Override
+        @NonNull
+        public DataSource<String, DataSnapshot> create() {
+            return new FirebaseDataSource(mQuery);
+        }
+    }
+
+    FirebaseDataSource(Query mQuery){
+        this.mQuery = mQuery;
+    }
+
+    @Override
+    public void loadInitial(@NonNull final LoadInitialParams<String> params,
+                            @NonNull final LoadInitialCallback<String, DataSnapshot> callback) {
+
+        // Set initial loading state
+        mLoadingState.postValue(LoadingState.LOADING_INITIAL);
+
+        Query mInitQuery = mQuery.limitToFirst(params.requestedLoadSize);
+        mInitQuery.addListenerForSingleValueEvent(new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
+                if (dataSnapshot.exists()) {
+
+                    //Make List of DataSnapshot
+                    List<DataSnapshot> data = new ArrayList<>();
+
+                    for (DataSnapshot snapshot : dataSnapshot.getChildren()){
+                        data.add(snapshot);
+                    }
+
+                    //Get Last Key
+                    String lastKey = getLastPageKey(data);
+
+                    //Update State
+                    mLoadingState.postValue(LoadingState.LOADED);
+                    mRetryRunnable = null;
+
+                    callback.onResult(data, lastKey, lastKey);
+
+                } else {
+                    mRetryRunnable = getRetryLoadInitial(params, callback);
+                    setDatabaseNotFoundError();
+                }
+            }
+
+            @Override
+            public void onCancelled(@NonNull DatabaseError databaseError) {
+                mRetryRunnable = getRetryLoadInitial(params, callback);
+                setError(databaseError);
+            }
+        });
+    }
+
+    @Override
+    public void loadBefore(@NonNull LoadParams<String> params, @NonNull LoadCallback<String, DataSnapshot> callback) {
+        // Ignored for now, since we only ever append to the initial load.
+    }
+
+    @Override
+    public void loadAfter(@NonNull final LoadParams<String> params,
+                          @NonNull final LoadCallback<String, DataSnapshot> callback) {
+
+        // Set loading state
+        mLoadingState.postValue(LoadingState.LOADING_MORE);
+
+        //Load params.requestedLoadSize+1 because, first data item is getting ignored.
+        Query mNewQuery = mQuery.startAt(null, params.key).limitToFirst(params.requestedLoadSize + 1);
```

Comment:
Is there a reason we use the strategy of calling `startAt` and then skipping the first item instead of calling `startAfter`?

Reply:
Actually, `Query` of *Firebase Realtime Database* is not having method like `startAfter()` to get data after skipping any value.
Also, `DataSnapshot` not have any method to convert it into `List<DataSnapshot>` so that we can pass its sublist from 1st index.
This is the reason here we used `startAt`

Reply:
D'oh you're right!  I was looking at Firestore.  My bad.